File: Stack_Queue/LargestRectInHistogram.cpp

/*
 * Author: Amitrajit Bose
 * Problem Link: https://www.spoj.com/problems/HISTOGRA/ 
 * Approach: In a empty stack keep adding all indices of increasing bars,
  if bar is smaller than bar at top of stack, then keep popping from stack
  while TOS is greater. Calculate area with the removed bar as smallest
  bar. In this case, the formula varies of stack is empty. After 
  traversing entire histogram, keep popping the stack and calcalate
  area to check if a larger rectangle can be found.
 */

#include<bits/stdc++.h>
#include <iostream>
#include<ctype.h>
#include<climits>
using namespace std;
typedef long long ll;

#define rep(i,m,n) for(i=m; i<(n); i++)
#define test ll t; cin>>t; while(t--)

ll maxArea(ll hist[], ll n)
{
	stack<ll> st;
	st.push(0); //push first index to stack
	ll maxm=0; //maximum area
	ll i,area,p;
	rep(i,1,n)
	{
		if(st.empty() or hist[i]>=hist[st.top()])
			st.push(i);
		else
		{
			//the bar is now sloping down
			while(!st.empty() and hist[i]<hist[st.top()])
			{
				p = st.top(); //kept the topmost item of stack in hand and pop
				st.pop();
				if(st.empty())
					area = hist[p]*i;
				else
					area = hist[p]*(i-st.top()-1);
				maxm=max(maxm,area);
			}
			st.push(i);//taken into consideration now
		}
	}
	// traversed the entire histogram
	// now pop the stack and recalculate for higher area, if any
	while(!st.empty())
	{
		p=st.top();
		st.pop();
		if(st.empty())
			area=hist[p]*i;
		else
			area=hist[p]*(i-st.top()-1);
		maxm=max(maxm,area);
	}
	//cout << maxm << endl;
	return maxm;
}

int main()
{
	ll n,i;
	cin >> n; //first n
	while(n!=0)
	{
		ll arr[n];
		rep(i,0,n)
			cin >> arr[i];
		// inputs taken
		cout << maxArea(arr, n) << endl;
		cin >> n; //next n
	}
	return 0;
}

File: Stack_Queue/LargestRectInHistogram.java
public class Solution {
    public int largestRectangleArea(int[] A) {
        
        Stack<Integer> st = new Stack<>();
        int maxarea = 0;
        int n = A.length;
        int lastElementIdx;
        
        if(n == 0) return 0;
        if(n == 1) return A[0];
        
        for (int i=0; i<n; i++){
            // System.out.println(st);
            while (!st.isEmpty() && A[st.peek()] > A[i]){
                lastElementIdx = st.pop();
                maxarea = Math.max(maxarea, A[lastElementIdx]*(i-(st.isEmpty() ? -1 : st.peek())-1));
            }
            st.push(i);
        }
        // since all elements have been processed
        // checking if the stack is empty or any unprocessed element is left
        while(!st.isEmpty()){
            lastElementIdx = st.pop();
            maxarea = Math.max(maxarea, A[lastElementIdx]*(n-(st.isEmpty() ? -1 : st.peek())-1));
        }
        return maxarea;
    }
}

